app/lib/dartdoc/dartdoc_page.dart

```diff
@@ -0,0 +1,419 @@
+// Copyright (c) 2022, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:collection/collection.dart';
+import 'package:html/dom.dart' show Element;
+import 'package:html/parser.dart' as html_parser;
+import 'package:pub_dev/frontend/dom/dom.dart' as d;
+import 'package:pub_dev/frontend/static_files.dart';
+import 'package:sanitize_html/sanitize_html.dart';
+
+import '../shared/urls.dart';
+
+class DartDocPageOptions {
+  final String package;
+  final String version;
+  final bool isLatestStable;
+
+  /// Path of the current file relative to the documentation root.
+  final String path;
+
+  DartDocPageOptions({
+    required this.package,
+    required this.version,
+    required this.isLatestStable,
+    required this.path,
+  });
+
+  String get latestStableDocumentationUrl => pkgDocUrl(package, isLatest: true);
+
+  String get pubPackagePageUrl =>
+      pkgPageUrl(package, version: isLatestStable ? null : version);
+
+  String get canonicalUrl {
+    var p = path;
+    // Strip /index.html
+    if (p.endsWith('/index.html')) {
+      p = p.substring(0, p.length - 'index.html'.length);
+    }
+    return isLatestStable
+        ? pkgDocUrl(
+            package,
+            includeHost: true,
+            isLatest: true,
+            relativePath: p,
+          )
+        : pkgDocUrl(
+            package,
+            includeHost: true,
+            version: version,
+            relativePath: p,
+          );
+  }
+}
+
+class Breadcrumb {
+  final String title;
+
+  /// Link to the breadcrumb, `null` if this the current page
+  final String? href;
+
+  Breadcrumb({required this.title, required this.href});
+}
+
+/// A dartdoc page parsed as sanitized html.
+///
+///
+/// This assumes a page layout along the lines of:
+/// ```
+///           [header]
+/// ------------------------------
+///        |           |
+/// [left] | [content] | [right]
+///        |           |
+/// ------------------------------
+///           [footer]
+/// ```
+class DartDocPage {
+  final String title;
+  final String description;
+  final List<Breadcrumb> breadcrumbs;
+  final String left;
```

Comment:
Let's add a note that these are HTML strings (in contrast with the other fields like title or description).